crates/bevy_render/src/render_resource/pipeline_cache.rs

```diff
@@ -325,220 +210,115 @@ impl LayoutCache {
-            PipelineDescriptor::ComputePipelineDescriptor(descriptor) => descriptor,
+    fn get_pipeline(&self, id: I) -> Option<&P> {
+        match &self.pipelines[id.index()].state {
+            PipelineState::Ok(pipeline) => Some(pipeline),
+            _ => None,
         }
     }
 
-    /// Try to retrieve a render pipeline GPU object from a cached ID.
-    ///
-    /// # Returns
-    ///
-    /// This method returns a successfully created render pipeline if any, or `None` if the pipeline
-    /// was not created yet or if there was an error during creation. You can check the actual creation
-    /// state with [`PipelineCache::get_render_pipeline_state()`].
     #[inline]
-    pub fn get_render_pipeline(&self, id: CachedRenderPipelineId) -> Option<&RenderPipeline> {
-        if let CachedPipelineState::Ok(Pipeline::RenderPipeline(pipeline)) =
-            &self.pipelines[id.0].state
-        {
-            Some(pipeline)
-        } else {
-            None
-        }
+    fn reset_pipeline(&mut self, id: I) {
+        self.pipelines[id.index()].state = PipelineState::Queued;
+        self.waiting_pipelines.insert(id);
     }
 
-    /// Try to retrieve a compute pipeline GPU object from a cached ID.
-    ///
-    /// # Returns
-    ///
-    /// This method returns a successfully created compute pipeline if any, or `None` if the pipeline
-    /// was not created yet or if there was an error during creation. You can check the actual creation
-    /// state with [`PipelineCache::get_compute_pipeline_state()`].
     #[inline]
-    pub fn get_compute_pipeline(&self, id: CachedComputePipelineId) -> Option<&ComputePipeline> {
-        if let CachedPipelineState::Ok(Pipeline::ComputePipeline(pipeline)) =
-            &self.pipelines[id.0].state
-        {
-            Some(pipeline)
-        } else {
-            None
-        }
-    }
-
-    /// Insert a render pipeline into the cache, and queue its creation.
-    ///
-    /// The pipeline is always inserted and queued for creation. There is no attempt to deduplicate it with
-    /// an already cached pipeline.
-    ///
-    /// # Returns
-    ///
-    /// This method returns the unique render shader ID of the cached pipeline, which can be used to query
-    /// the caching state with [`get_render_pipeline_state()`] and to retrieve the created GPU pipeline once
-    /// it's ready with [`get_render_pipeline()`].
-    ///
-    /// [`get_render_pipeline_state()`]: PipelineCache::get_render_pipeline_state
-    /// [`get_render_pipeline()`]: PipelineCache::get_render_pipeline
-    pub fn queue_render_pipeline(
-        &self,
-        descriptor: RenderPipelineDescriptor,
-    ) -> CachedRenderPipelineId {
+    fn queue_pipeline(&self, descriptor: D) -> I {
         let mut new_pipelines = self.new_pipelines.lock();
-        let id = CachedRenderPipelineId(self.pipelines.len() + new_pipelines.len());
+        let id = I::new();
```

Comment:
This assumes the atomic counter stays in sync with the growing size of the Vec, which is not necessarily true if someone makes a second `PipelineCache`.

Reply:
Can we enforce that we use a single PipelineCache?
Otherwise, I could also special case both Ids and add the counter to the `PipelineCacheInternal` instead.

Reply:
If you had multiple `PipelineCache` instances, how would you know from which to look stuff up?

Reply:
First of all, why would there be multiple `PipelineCache`s? It is a resource, so it should be exclusive, right?

Technically you can create and use multiple caches if you wanted. So we might want to prevent that behavior, or move the id counter into the pipeline cache. 
But IDs created by multiple caches could then be mistaken as the same ID.
To prevent this we could also have an additional cache counter and store on each ID which cache it belongs to.

e.g. First 8 bits of each ID for the cache and the other 24 for the pipeline should be more than enough anyways.

This all sounds pretty hypothetical and unrealistic to me though.

What do you think? Does simple documentation of this behavior suffice?

Reply:
I have special-cased both Ids. Each pipeline cache now has its own atomic counter, that increments the ids.

There is still the issue of ID collisions between two pipelines, but I doubt that this will be a real problem.

Reply:
> First of all, why would there be multiple `PipelineCache`s? It is a resource, so it should be exclusive, right?

There may be multiple Worlds, and thus multiple resources that are being allocated for, and the constructor is public.
 
ID collisions between caches is something I don't think we should be defending against. There's already precendence for container-specific ID schemes in Bevy already (Entities in World, for example).

Reply:
Ok, so I will simply add another atomic counter, which counts the caches and add this to the ids?

Reply:
No I think the internal counter is fine. Might not even need it to be atomic either if we just take the `current.len() + waiting.len() + 1`.

Reply:
Ok, I think the latest change should be it then.